beacon-chain/rpc/beacon_server.go

```diff
@@ -136,6 +137,71 @@ func (bs *BeaconServer) BlockTree(ctx context.Context, _ *ptypes.Empty) (*pb.Blo
 	}, nil
 }
 
+// eth1Data takes all eth1data votes validate the relevant ones,
+// checks that they are valid against eth1 chain data, count them
+// and returns the best vote.
+func (bs *BeaconServer) eth1Data(ctx context.Context) (*pbp2p.Eth1Data, error) {
+	beaconState, err := bs.beaconDB.HeadState(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch beacon state: %v", err)
+	}
+	currentHeight := bs.powChainService.LatestBlockHeight()
+	eth1FollowDistance := int64(params.BeaconConfig().Eth1FollowDistance)
+	stateLatestEth1Hash := bytesutil.ToBytes32(beaconState.LatestEth1Data.DepositRoot)
+	if stateLatestEth1Hash == [32]byte{} {
+		return bs.defaultEth1DataResponse(ctx, currentHeight, eth1FollowDistance)
+	}
+	// Fetch the height of the block pointed to by the beacon state's latest_eth1_data.block_hash
+	// in the canonical eth1.0 chain.
+	_, stateLatestEth1Height, err := bs.powChainService.BlockExists(ctx, stateLatestEth1Hash)
+	if err != nil {
+		return nil, fmt.Errorf("could not verify block with hash exists in Eth1 chain: %#x: %v", stateLatestEth1Hash, err)
+	}
+	dataVotes := []*pbp2p.Eth1Data{}
+	votesMap := helpers.EmptyVoteHierarchyMap()
+	depositCount, depositRootAtHeight := bs.beaconDB.DepositsNumberAndRootAtHeight(ctx, currentHeight)
+	for _, vote := range beaconState.Eth1DataVotes {
+		validVote, blockHeight, err := bs.validateVote(ctx, currentHeight, depositCount, depositRootAtHeight, eth1FollowDistance, stateLatestEth1Height, vote)
+		if err != nil {
+			return nil, err
+		}
+		if validVote {
+			dataVotes = append(dataVotes, vote)
+			votesMap, err = helpers.CountVote(votesMap, vote, blockHeight)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	if len(dataVotes) == 0 {
+		return bs.defaultEth1DataResponse(ctx, currentHeight, eth1FollowDistance)
+	}
+	return votesMap.BestVote, nil
+
+}
+
+func (bs *BeaconServer) validateVote(ctx context.Context, currentHeight *big.Int, depositCount uint64, depositRootAtHeight [32]byte, eth1FollowDistance int64, stateLatestEth1Height *big.Int, vote *pbp2p.Eth1Data) (bool, *big.Int, error) {
```

Comment:
comment would be nice too

Reply:
Why does validate vote return a block height?

Reply:
Why is eth1 follow distance a parameter? isn't it a config constant?

Reply:
removed